### DISCUSSION TITLE
Go module retrieval issue for private gitlab instances not hosted at the top level

### DISCUSSION THREAD

**Discussion by btajuddin-planet** [renovatebot/renovate#23194] [Request Help]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us what version of Renovate you run.

docker tag 35.144.0

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### Was this something which used to work for you, and then stopped?

I am trying to get this working for the first time

### Describe the problem

This is a somewhat complicated issue, so I'll start with a couple facts of our setup:
1. go.example.com is the base domain for all of our go modules and the value of `GOPRIVATE`
2. The modules' source code lives in a self-hosted gitlab instance at https://internal.example.com/code/
3. The domain in point 1 properly points to the repositories in the gitlab in point 2.
4. Renovate is running inside a CI/CD job in the gitlab from point 2.

For a pretend package `foo/bar`, the [Go datasource](https://github.com/renovatebot/renovate/blob/main/lib/modules/datasource/go/base.ts) will make a call to `https://go.example.com/product/mono` to get the location of the source. The logs demonstrate that the import URL was determined to be `https://internal.example.com/code/product/mono.git`.

So far, so good.

The next thing the logs show is a failing call to `https://internal.example.com/api/v4/projects/code%2Fproduct%2Fmono/repository/tags?per_page=100`. This is not good or correct. The `code` slug on the URL has been moved inside the repository path.

Ultimately, I believe the problem occurs at [this line](https://github.com/renovatebot/renovate/blob/0bcd387b89f866172f43aa536b1ee954502280dc/lib/modules/datasource/go/base.ts#L148), but a fix would ripple beyond that.

I've looked through a lot of configs and documentation without an idea of how to bypass this issue. It would be nice to have it fall back onto a basic git tags behavior if this doesn't work (since that should), but instead it just fails and keeps going with other packages.

As you can see, it would be very difficult to create a reproducible use case for this since it would require having a gitlab instance at a non-host-level URL exposed to the public.

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: Go module: go.example.com/product/mono lookup import url https://internal.example.com/code/product/mono.git (repository=product/golib)
DEBUG: GET https://internal.example.com/api/v4/projects/code%2Fproduct%2Fmono/repository/tags?per_page=100 = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=404 retryCount=0, duration=484) (repository=product/golib)
DEBUG: GitLab API 404 (repository=product/golib)
       "url": "https://internal.example.com/api/v4/projects/code%2Fproduct%2Fmono/repository/tags?per_page=100"
DEBUG: Datasource 404 (repository=product/golib)
       "datasource": "go",
       "packageName": "go.example.com/product/mono",
       "url": "https://internal.example.com/api/v4/projects/code%2Fproduct%2Fmono/repository/tags?per_page=100"
DEBUG: Failed to look up go package go.example.com/product/mono (repository=product/golib, packageFile=v2/go.mod, dependency=go.example.com/product/mono)
```

</details>

### Have you created a minimal reproduction repository?

I have explained in the description why a minimal reproduction is impossible

**Comment by rarkins** [comment]
So this problem applies to self-hosted GitLab instances which use a subdirectory (e.g. `/code`)?

**Comment by btajuddin-planet** [comment]
Yes. It appears to be specific to that sort of configuration.

**Reply by rarkins** [reply]
I'm not sure if there's any easy way to solve this. Brainstorming:
- Can Renovate somehow "learn" that `https://internal.example.com/code/` is the base url and not `https://internal.example.com/`?
- How does Renovate determine that it's a GitLab host in the first place?
- Should we be using raw git instead of gitlab tags?

**Reply by btajuddin-planet** [reply]
Agreed that there is no easy fix. To your points:

- I wouldn't require that renovate "learn" so much as allow me to tell it. I don't know if there's a substantive difference with what you're thinking. I understand that this case is probably a little bit niche and don't expect an automated solution.
- We are running renovate on the gitlab instance hosted in the subdirectory. While we don't have a host rule defined for it, might there be one automatically defined for it?
- This is, I think, the most likely path forward. I suspect that using raw git is slower and probably involves a clone, where github/gitlab/etc. APIs are simpler and faster. That said, it seems like a reasonable fallback for any failed tag lookup of a known git repo. Sure, try the API first, but if it doesn't work, we can definitely try standard git semantics, since we already know it's a git repo.

**Comment by rarkins** [comment]
> * We are running renovate on the gitlab instance hosted in the subdirectory. While we don't have a host rule defined for it, might there be one automatically defined for it?

Do you mean that your `endpoint` in Renovate config is `https://internal.example.com/code/`?

**Reply by btajuddin-planet** [reply]
The endpoint is `https://internal.example.com/code/api/v4` based on the template linked from the docs.

**Reply by rarkins** [reply]
OK so in theory Renovate could have a special rule to reason about that and figure out that `/code/` is part of the prefix, not the module

**Reply by viceice** [reply]
I think we should message `endpoint` to not include the api . then we can use that to check prefix

**Reply by btajuddin-planet** [reply]
That's a possibility that would solve my particular use case. It isn't a general solution, though. If someone decided to create a setup like this that is open to the public, the self-hosted gitlab URL wouldn't be the endpoint in the renovate config. I think it's still important to be able to either define a rule to establish the base URL of a self-hosted instance or fall back to vanilla git tags.

**Reply by viceice** [reply]
right, I've looked at the code and it seems we can use the matchHost from the host rules to detect the prefix.

@rarkins WDYT?
